tests/test.py

```diff
@@ -262,79 +293,190 @@ def read_write_history_pipe():
-        "phys": "input-remapper/input1",
-        "info": evdev.device.DeviceInfo(5, 1, 5, 1),
-        "name": "input-remapper Bar Device",
-    },
+    dev_input_event40 = Fixture(
+        capabilities={evdev.ecodes.EV_KEY: keyboard_keys},
+        phys="input-remapper/input1",
+        info=evdev.device.DeviceInfo(5, 1, 5, 1),
+        name="input-remapper Bar Device",
+        path="/dev/input/event40",
+    )
     # denylisted
-    "/dev/input/event51": {
-        "capabilities": {evdev.ecodes.EV_KEY: keyboard_keys},
-        "phys": "usb-0000:03:00.0-5/input1",
-        "info": evdev.device.DeviceInfo(6, 1, 6, 1),
-        "name": "YuBiCofooYuBiKeYbar",
-    },
-}
+    dev_input_event51 = Fixture(
+        capabilities={evdev.ecodes.EV_KEY: keyboard_keys},
+        phys="usb-0000:03:00.0-5/input1",
+        info=evdev.device.DeviceInfo(6, 1, 6, 1),
+        name="YuBiCofooYuBiKeYbar",
+        path="/dev/input/event51",
+    )
 
+    def __init__(self):
+        self._iter = [
+            self.dev_input_event1,
+            self.dev_input_event11,
+            self.dev_input_event10,
+            self.dev_input_event13,
+            self.dev_input_event14,
+            self.dev_input_event15,
+            self.dev_input_event20,
+            self.dev_input_event30,
+            self.dev_input_event31,
+            self.dev_input_event40,
+            self.dev_input_event51,
+        ]
+        self._dynamic_fixtures = {}
+
+    def __getitem__(self, item: str) -> Fixture:
+        if fixture := self._dynamic_fixtures.get(item):
+            return fixture
+        item = self._path_to_attribute(item)
 
-def setup_pipe(group_key):
+        try:
+            return getattr(self, item)
+        except AttributeError as e:
+            raise KeyError(str(e))
+
+    def __setitem__(self, key: str, value: [Fixture | dict]):
+        if isinstance(value, Fixture):
+            self._dynamic_fixtures[key] = value
+        elif isinstance(value, dict):
+            self._dynamic_fixtures[key] = Fixture(path=key, **value)
+
+    def __iter__(self):
+        return iter([*self._iter, *self._dynamic_fixtures.values()])
+
+    def reset(self):
+        self._dynamic_fixtures = {}
+
+    @staticmethod
+    def _path_to_attribute(path) -> str:
+        if path.startswith("/"):
+            path = path[1:]
+        if "/" in path:
+            path = path.replace("/", "_")
+        return path
+
+    def get(self, item) -> Optional[Fixture]:
+        try:
+            return self[item]
+        except KeyError:
+            return None
+
+    @property
+    def foo_device_1_1(self):
+        return self["/dev/input/event1"]
+
+    @property
+    def foo_device_2_mouse(self):
+        return self["/dev/input/event11"]
+
+    @property
+    def foo_device_2_keyboard(self):
+        return self["/dev/input/event10"]
+
+    @property
+    def foo_device_2_13(self):
+        return self["/dev/input/event13"]
+
+    @property
+    def foo_device_2_qux(self):
```

Comment:
is `self.foo_device_2_qux` the same as `self.dev_input_event14`?

Reply:
yes

Reply:
isn't it redundant then?

Reply:
yes it is. I created the property for convenience to improve the readability. The attribute name `dev_input_eventXX` allows us to access the attribute with `fixtures["/dev/input/eventXX"]` from the `__getitem__` method, which is necessary for backwards comparability with some tests and for dynamic fixtures.

Reply:
ok